biospecimen-tracking/api/src/main/java/org/phenotips/data/internal/controller/BiospecimensController.java

```diff
@@ -0,0 +1,379 @@
+                if (biospecimen != null) {
+                    result.add(biospecimen);
+                }
+            }
+        } catch (ParseException e) {
+            this.logger.error("Unable to parse JSON data [{}]", e.getMessage());
+            return null;
+        }
+
+        return new IndexedPatientData<>(getName(), result);
+    }
+
+    /**
+     * Gets an EntityReference to the xWiki class representing BiospecimenData in the patient document.
+     * @return a EntityReference object
+     */
+    public static EntityReference getClassReference()
+    {
+        return new EntityReference(CLASS_NAME, EntityType.DOCUMENT, Constants.CODE_SPACE_REFERENCE);
+    }
+
+    /***************************************** Private Methods ****************************************/
+
+    /**
+     * Helper function. If selectedFieldNames is no null it will be returned, otherwise the PROPERTIES set will be
+     * returned otherwise.
+     *
+     * @param selectedFieldNames a collection of field names to return if not null
+     * @return a Collection
+     */
+    private static Collection<String> propertyNamesToIterateOver(Collection<String> selectedFieldNames)
+    {
+        if (CollectionUtils.isEmpty(selectedFieldNames)) {
+            return BiospecimenData.PROPERTIES;
+        } else {
+            return selectedFieldNames;
+        }
+    }
+
+    private static BiospecimenData parseBiospecimenObject(Object biospecimenObject, DateFormat jsonDateFormat)
+        throws ParseException {
+
+
+        if (!(biospecimenObject instanceof JSONObject)) {
+            return null;
+        }
+
+        JSONObject biospecimenJSONObject = (JSONObject) biospecimenObject;
+
+        BiospecimenData biospecimen = new BiospecimenData();
+
+        for (String property : BiospecimenData.PROPERTIES) {
+            if (!biospecimenJSONObject.has(property)) {
+                continue;
+            }
+
+            Object propertyObject = biospecimenJSONObject.get(property);
+
+            if (propertyObject == null) {
+                continue;
+            }
+
+            switch (property) {
+                case BiospecimenData.TYPE_PROPERTY_NAME:
+                    biospecimen.setType(String.valueOf(propertyObject));
+                    break;
+                case BiospecimenData.DATE_COLLECTED_PROPERTY_NAME:
+                    biospecimen.setDateCollected(jsonDateFormat.parse(String.valueOf(propertyObject)));
+                    break;
+                case BiospecimenData.DATE_RECEIVED_PROPERTY_NAME:
+                    biospecimen.setDateReceived(jsonDateFormat.parse(String.valueOf(propertyObject)));
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        return biospecimen;
+    }
+
+    private static JSONObject getJSONObjectFromBiospecimenData(BiospecimenData biospecimen,
```

Comment:
50/50, can you give me a compelling reason?

I'm not convinced since it is not really a list of possible/allowed values for a strong-typed data type.